cluster-autoscaler/core/scale_up.go

```diff
@@ -283,32 +278,49 @@ func ScaleUp(context *context.AutoscalingContext, processors *ca_processors.Auto
 		scaleUpInfos, typedErr := nodegroupset.BalanceScaleUpBetweenGroups(
 			targetNodeGroups, newNodes)
 		if typedErr != nil {
-			return false, typedErr
+			return false, &status.ScaleUpStatus{}, typedErr
 		}
 		glog.V(1).Infof("Final scale-up plan: %v", scaleUpInfos)
 		for _, info := range scaleUpInfos {
 			typedErr := executeScaleUp(context, clusterStateRegistry, info)
 			if typedErr != nil {
-				return false, typedErr
+				return false, &status.ScaleUpStatus{}, typedErr
 			}
 		}
 
-		for _, pod := range bestOption.Pods {
-			context.Recorder.Eventf(pod, apiv1.EventTypeNormal, "TriggeredScaleUp",
-				"pod triggered scale-up: %v", scaleUpInfos)
-		}
-
 		clusterStateRegistry.Recalculate()
-		return true, nil
+		return true, &status.ScaleUpStatus{
+				ScaleUpInfos:            scaleUpInfos,
+				PodsRemainUnschedulable: podSetToList(podsRemainUnschedulable),
+				PodsTriggeredScaleUp:    bestOption.Pods,
+				PodsAwaitEvaluation:     getPodsAwaitingEvaluation(unschedulablePods, podsRemainUnschedulable, bestOption.Pods)},
+			nil
 	}
-	for pod, unschedulable := range podsRemainUnschedulable {
-		if unschedulable {
-			context.Recorder.Event(pod, apiv1.EventTypeNormal, "NotTriggerScaleUp",
-				"pod didn't trigger scale-up (it wouldn't fit if a new node is added)")
+
+	return false, &status.ScaleUpStatus{PodsRemainUnschedulable: podSetToList(podsRemainUnschedulable)}, nil
+}
+
+func podSetToList(podSet map[*apiv1.Pod]bool) []*apiv1.Pod {
```

Comment:
This doesn't just convert a map to a list, but also filters out keys for which the value is false. How about "getRemainingPods" or sth like this?

Reply:
That's why it's not podMap but podSet, if you treat the map as a set (and having a pod to bool map is an ok option for a set in go AFAIU). However, if it's not immediately clear, I guess I should change it.